cmd/pebble/cmd_enter.go

```diff
@@ -0,0 +1,186 @@
+manager is shut down after the commands finishes unless noted otherwise below.
+
+Only a subset of pebble commands is supported as a subcommand:
+
+  exec
+  help
+  ls
+  plan
+  services
+  start
+  version
+
+Furthermore, some commands have restrictions and idiosyncrasies:
+
+  exec, ls, plan, services, version
+    Service manager logging is disabled unless -v/--verbose flag is provided.
+
+  help, ls, plan, services, version
+    Service autostart (the --run flag) is prohobitied.
+
+  start, stop
+    Service manager continues running after the command finishes.
+
+  stop
+    Service autostart (the --run flag) is required.
+`
+
+type enterFlags int
+
+const (
+	enterFlagSilenceLogging enterFlags = 1 << iota
+	enterFlagNoServiceManager
+	enterFlagKeepServiceManager
+	enterFlagRequireServiceAutostart
+	enterFlagProhibitServiceAutostart
+)
+
+type cmdEnter struct {
+	clientMixin
+
+	sharedRunEnterOpts
+	Run        bool `long:"run"`
+	Positional struct {
+		Cmd []string `positional-arg-name:"<subcommand>"`
+	} `positional-args:"yes"`
+}
+
+func init() {
+	optsHelp := map[string]string{
+		"run": "Start \"autostart\" services before executing subcommand",
+	}
+	for k, v := range sharedRunEnterOptsHelp {
+		optsHelp[k] = v
+	}
+	cmdInfo := addCommand("enter", shortEnterHelp, longEnterHelp, func() flags.Commander { return &cmdEnter{} }, optsHelp, nil)
+	cmdInfo.extra = func(cmd *flags.Command) {
+		cmd.PassAfterNonOption = true
+	}
+}
+
+func commandEnterFlags(commander flags.Commander) (enterFlags enterFlags, supported bool) {
+	supported = true
+	switch commander.(type) {
+	case *cmdExec:
+		enterFlags = enterFlagSilenceLogging
+	case *cmdHelp:
+		enterFlags = enterFlagNoServiceManager
+	case *cmdLs, *cmdPlan, *cmdServices, *cmdVersion:
+		enterFlags = enterFlagSilenceLogging | enterFlagProhibitServiceAutostart
+	case *cmdStart:
+		enterFlags = enterFlagKeepServiceManager
+	case *cmdStop:
+		enterFlags = enterFlagKeepServiceManager | enterFlagRequireServiceAutostart
+	default:
+		supported = false
+	}
+	return
+}
+
+func (cmd *cmdEnter) Execute(args []string) error {
+	if len(args) > 0 {
+		return ErrExtraArgs
+	}
+
+	runCmd := cmdRun{
+		sharedRunEnterOpts: cmd.sharedRunEnterOpts,
+	}
+	runCmd.setClient(cmd.client)
+
+	if len(cmd.Positional.Cmd) == 0 {
+		runCmd.run(nil)
+		return nil
+	}
+
+	runCmd.Hold = !cmd.Run
+
+	var (
+		commander flags.Commander
+		extraArgs []string
+	)
+
+	parser := Parser(cmd.client)
+	parser.CommandHandler = func(c flags.Commander, a []string) error {
+		commander = c
+		extraArgs = a
+		return nil
+	}
+
+	if _, err := parser.ParseArgs(cmd.Positional.Cmd); err != nil {
+		return err
+	}
+
+	enterFlags, supported := commandEnterFlags(commander)
+
+	if !supported {
+		return fmt.Errorf("enter: this command is not supported")
+	}
+
+	if enterFlags&enterFlagRequireServiceAutostart != 0 && !cmd.Run {
+		return fmt.Errorf("enter: this command can only run with autostarted services (hint: add --run to enter)")
```

Comment:
Perhaps more directly:

```
enter: must use --run before %q subcommand
```

With that said, the memory I have of our final agreement (which arguably took a bit of back and forth, so I may easily be missing something) was that we'd make subcommands behave in their most natural way, and when there was no natural way at all we'd prevent the subcommand from running.

In particular here, the fact we'd have _start_ and _stop_ disagreeing on the requirement of _--run_, and the fact we're forcing _--run_ to be provided to a subcommand that could not work without it, are both bad symptoms that this isn't quite what we need. Let's talk about it elsewhere to revisit this topic please.

Reply:
I wholeheartedly agree that the sub-command behavior as currently specified is confusing and hard to convey.